File: Palace/Settings/TPPSettingsSplitViewController.m
#import "TPPSettingsPrimaryNavigationController.h"
#import "TPPSettingsPrimaryTableViewController.h"
#import "TPPSettingsEULAViewController.h"
#import "TPPBook.h"
#import "TPPMyBooksDownloadCenter.h"
#import "TPPRootTabBarController.h"
#import "Palace-Swift.h"

#import "TPPSettingsSplitViewController.h"

@interface TPPSettingsSplitViewController ()
  <UISplitViewControllerDelegate, NYPLSettingsPrimaryTableViewControllerDelegate>

@property (nonatomic) TPPSettingsPrimaryNavigationController *primaryNavigationController;
@property (nonatomic) bool isFirstLoad;
@property (nonatomic) id<TPPCurrentLibraryAccountProvider> currentLibraryAccountProvider;
@end

@implementation TPPSettingsSplitViewController

#pragma mark NSObject

- (instancetype)initWithCurrentLibraryAccountProvider: (id<TPPCurrentLibraryAccountProvider>)currentAccountProvider
{
  self = [super init];
  if(!self) return nil;
  
  self.delegate = self;
  self.currentLibraryAccountProvider = currentAccountProvider;

  self.title = NSLocalizedString(@"Settings", nil);
  self.tabBarItem.image = [UIImage imageNamed:@"Settings"];
  
  self.primaryNavigationController = [[TPPSettingsPrimaryNavigationController alloc] init];
  self.primaryNavigationController.primaryTableViewController.delegate = self;
  self.viewControllers = @[self.primaryNavigationController];
  
  self.presentsWithGesture = NO;
  
  return self;
}

- (void)viewDidLoad
{
  [super viewDidLoad];
  
  NSArray *accounts = [[TPPSettings sharedSettings] settingsAccountsList];
  
  if (UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPad &&
     (self.traitCollection.horizontalSizeClass != UIUserInterfaceSizeClassCompact)) {
    
    self.viewControllers = @[self.primaryNavigationController,
                             [[UINavigationController alloc] initWithRootViewController:
                              [[TPPSettingsAccountsTableViewController alloc] initWithAccounts:accounts]]];
    
    [self highlightFirstTableViewRow:YES];
  } else {
    self.viewControllers = @[self.primaryNavigationController];
  }
  
  self.preferredDisplayMode = UISplitViewControllerDisplayModeAllVisible;
  self.isFirstLoad = YES;
}

- (void)highlightFirstTableViewRow:(bool)highlight
{
  if (highlight) {
    [self.primaryNavigationController.primaryTableViewController.tableView
     selectRowAtIndexPath:NYPLSettingsPrimaryTableViewControllerIndexPathFromSettingsItem(NYPLSettingsPrimaryTableViewControllerItemAccount)
     animated:NO
     scrollPosition:UITableViewScrollPositionMiddle];
  } else {
    [self.primaryNavigationController.primaryTableViewController.tableView
     deselectRowAtIndexPath:NYPLSettingsPrimaryTableViewControllerIndexPathFromSettingsItem(                                     NYPLSettingsPrimaryTableViewControllerItemAccount)
     animated:NO];
  }
}

#pragma mark UISplitViewControllerDelegate

- (BOOL)splitViewController:(__attribute__((unused)) UISplitViewController *)splitViewController
collapseSecondaryViewController:(__attribute__((unused)) UIViewController *)secondaryViewController
ontoPrimaryViewController:(__attribute__((unused)) UIViewController *)primaryViewController
{
  if (self.isFirstLoad) {
    self.isFirstLoad = NO;
    return YES;
  } else {
    self.isFirstLoad = NO;
    return NO;
  }
}

#pragma mark NYPLSettingsPrimaryTableViewControllerDelegate

- (void)settingsPrimaryTableViewController:(__attribute__((unused)) TPPSettingsPrimaryTableViewController *const)settingsPrimaryTableVC
                             didSelectItem:(NYPLSettingsPrimaryTableViewControllerItem const)item
{
  UIViewController *viewController;
  NSArray *accounts;
  switch(item) {
    case NYPLSettingsPrimaryTableViewControllerItemAccount:
      accounts = [[TPPSettings sharedSettings] settingsAccountsList];
      viewController = [[TPPSettingsAccountsTableViewController alloc] initWithAccounts:accounts];
      break;
    case NYPLSettingsPrimaryTableViewControllerItemAbout:
      viewController = [[RemoteHTMLViewController alloc]
                        initWithURL:[NSURL URLWithString: TPPSettings.NYPLAboutSimplyEURLString]
                        title:NSLocalizedString(@"AboutApp", nil)
                        failureMessage:NSLocalizedString(@"The page could not load due to a connection error.", nil)];
      break;
    case NYPLSettingsPrimaryTableViewControllerItemEULA:
      viewController = [[RemoteHTMLViewController alloc]
                        initWithURL:[NSURL URLWithString: TPPSettings.NYPLUserAgreementURLString]
                        title:NSLocalizedString(@"EULA", nil)
                        failureMessage:NSLocalizedString(@"The page could not load due to a connection error.", nil)];
      break;
    case NYPLSettingsPrimaryTableViewControllerItemSoftwareLicenses:
      viewController = [[BundledHTMLViewController alloc]
                        initWithFileURL:[[NSBundle mainBundle]
                                         URLForResource:@"software-licenses"
                                         withExtension:@"html"]
                        title:NSLocalizedString(@"SoftwareLicenses", nil)];
      break;
    case NYPLSettingsPrimaryTableViewControllerItemDeveloperSettings:
      viewController = [[TPPDeveloperSettingsTableViewController alloc] init];
      break;
    default:
      return;
  }
  
  [self showDetailViewController:[[UINavigationController alloc]
                                  initWithRootViewController:viewController]
                          sender:self];
}

@end

File: Palace/Settings/TPPSettingsSplitViewController.swift
//
//  TPPSettingsSplitViewController.swift
//  The Palace Project / Open eBooks
//
//  Created by Kyle Sakai.
//  Copyright © 2020 NYPL Labs. All rights reserved.
//

/// Currently used only by Open eBooks, but extendable for use in SimplyE.
/// - Seealso: https://github.com/NYPL-Simplified/Simplified-iOS/pull/1070
/// TODO: SIMPLY-3053
class TPPSettingsSplitViewController : UISplitViewController, UISplitViewControllerDelegate {
  private var isFirstLoad: Bool
  private var currentLibraryAccountProvider: TPPCurrentLibraryAccountProvider
  
  @objc init(currentLibraryAccountProvider: TPPCurrentLibraryAccountProvider) {
    self.isFirstLoad = true
    self.currentLibraryAccountProvider = currentLibraryAccountProvider
    let navVC = UINavigationController.init(rootViewController: TPPSettingsPrimaryTableViewController())
    super.init(nibName: nil, bundle: nil)
    
    self.delegate = self
    self.title = DisplayStrings.Settings.settings
    self.tabBarItem.image = UIImage.init(named: "Settings")
    self.viewControllers = [navVC]
    self.presentsWithGesture = false
  }
  
  @available(*, unavailable)
  required init?(coder: NSCoder) {
    fatalError("init(coder:) has not been implemented")
  }
  
  var primaryTableVC: TPPSettingsPrimaryTableViewController? {
    let navVC = self.viewControllers.first as? UINavigationController
    return navVC?.viewControllers.first as? TPPSettingsPrimaryTableViewController
  }

  // MARK: UIViewController

  override func viewDidLoad() {
    super.viewDidLoad()
    self.preferredDisplayMode = .allVisible

    configPrimaryVCItems(using:
      TPPLibraryAccountURLsProvider(account:
        currentLibraryAccountProvider.currentAccount))
  }
  
  // MARK: UISplitViewControllerDelegate

  func splitViewController(_ splitVC: UISplitViewController,
                           collapseSecondary secondaryVC: UIViewController,
                           onto primaryVC: UIViewController) -> Bool {
    let rVal = self.isFirstLoad
    self.isFirstLoad = false
    return rVal
  }
}
